app/services/DriveService.scala

```diff
@@ -34,34 +35,90 @@ object DriveService {
       .setTeamDriveId(Config.teamDriveId)
 
     request.execute().getFiles.asScala.toList
-      .map(file => { // extract this into a pure function we can test
-
-        // raw model file
-        // function transforms
-        // output target models
-
-        // The inline Try will handle nulls returned from Java library
-        val customProps = Try {
-          file.getProperties.asScala.toMap
-        }.toOption.getOrElse(Map.empty)
-        val thumbnailLink = Try {
-          file.getThumbnailLink
-        }.toOption.getOrElse("")
-        val exportLinks = Try {
-          file.getExportLinks.asScala.toMap
-        }.toOption.getOrElse(Map.empty)
-
-        File(
-          id = file.getId,
-          title = file.getName,
-          output = file.getMimeType,
-          outputPreview = file.getWebViewLink, // TODO - returning editable links - need to process to preview
-          outputDownload = file.getWebContentLink, // TODO - returning nulls
-          outputThumbnail = thumbnailLink, // TODO - returning 404s
-          exportLinks = exportLinks,
-          customProperties = customProps
-        )
-      })
+      .map(file =>
+        // transformToResearchRecord(extractRawFile(file)) // define functions with def
+        (extractRawFile andThen transformToResearchRecord)(file) // define functions with val
+    )
+  }
+
+  val extractRawFile = (fileStream: File) => {
+    val customProps = Try {
+      fileStream.getProperties.asScala.toMap
+    }.toOption.getOrElse(Map.empty)
+    val exportLinks = Try {
+      fileStream.getExportLinks.asScala.toMap
+    }.toOption.getOrElse(Map.empty)
+
+    DriveFile(
+      id = fileStream.getId,
+      title = fileStream.getName,
+      output = fileStream.getMimeType,
+      outputPreview = fileStream.getWebViewLink,
+      outputDownload = fileStream.getWebContentLink, // TODO - returning nulls
+      exportLinks = exportLinks,
+      customProperties = customProps
+    )
+  }
+
+  val transformToResearchRecord = (file: DriveFile) => {
+    ResearchRecord.apply(
+      id = file.id,
+      title = file.title,
+      team = file.customProperties.getOrElse("team", "Unknown Team"),
+      output = setOutputType(file.output),
+      outputUrl = convertToPreviewLink(file.outputPreview),
+      outputThumbnail = createThumbnailLink(file.id),
+      outputDownload = file.outputDownload, // TODO returning null - permissions issue?
+      okr = file.customProperties.getOrElse("okr", ""),
+      year = file.customProperties.getOrElse("year", "0").toInt,
+      quarter = setQuarter(file.customProperties.getOrElse("quarter", "Unknown Quarter")),
+      tags = setRandomTags(file.customProperties.getOrElse("tags", "No tags found"))
+    )
+  }
+
+  def convertToPreviewLink(str: String) = {
+    str.replaceAll("edit", "preview")
```

Comment:
works I'm sure, but be aware that manipulation of untrusted strings is a quick way to obscure issues including security issues.  This is because this replaceAll will happily change more or less than you intend.  That's probably OK at this level, but once you start replacing quotes or backslashes things can get unreliable.